### DISCUSSION TITLE
Creating a complex visual search experiment

### DISCUSSION THREAD

**Discussion by Eslifkin23** [jspsych/jsPsych#953] [Q&A]
Hi,

I'm not sure of the best way to even ask this question, but I'll try.  I have created a complete visual search experiment in PsychoPy, but I need to re-create it in jsPsych (it's a long story as to why).  I have never coded anything previously (except minor things in PsychoPy) and I have never used jsPsych before.

Basically what I need is stimuli to be arranged on an invisible 10x8 grid and all stimuli will change every trial.  There will be between 0-3 target items on each trials and subjects will have to click on each target and then click a Done button to end the trial.

I have all the information I need in an Excel file (trial numbers, block numbers, number of targets for each trial, target items, distractor items, locations, etc.).  I know that jsPsych can't read Excel/CSV files the same way that PsychoPy and other programs can, so I'm not sure what to do with this information but this is exactly how the experiment needs to run (important).

The basic experiment structure will be instructions, 16 practice trials with feedback, then 5 blocks of 32 trials each.  I'm not sure creating a plugin for a block of trials will work for what I need, since each trial is unique.

Any suggestions on how to go about creating this or any resources I can use to help myself would be greatly appreciated!

-Elisabeth

**Comment by jodeleeuw** [comment]
Hi Elisabeth,

There's a lot to learn to accomplish this, but hopefully the resources that we have available and this forum will help you learn everything you need to do it.

I'd recommend starting with our standard tutorials: https://www.jspsych.org/tutorials/hello-world/ and https://www.jspsych.org/tutorials/rt-task/. 

We also recently held an [online workshop](www.movingresearchonline.info) and the [video recording for Day 1](https://www.youtube.com/watch?v=BuhfsIFRFe8&feature=youtu.be) is also a good starting point for folks new to jsPsych. Some of the content in that video is redundant with the written tutorials, so you could choose whether you prefer video or text based content.

If you like video-based tutorials, I've published a handful of [jsPsych tutorials on YouTube](https://www.youtube.com/playlist?list=PLnfo1lBY1P2Mf_o6rV5wiqqn92Mw3UTGh). The dichotic listening task video is aimed at newer users.

It sounds like you will need to create a custom plugin for your experiment. The plugin will serve as the template for the visual search task that you need to create. You can set up parameters within the plugin to control all the aspects of a trial that are unique to that trial. We talk about plugin creation in [Day 3](https://www.youtube.com/watch?v=LP7o0iAALik&feature=youtu.be) of our workshop. This will be the trickiest bit of setting up the experiment for a novice programmer. 

Please feel welcome to utilize this forum for specific questions as they come up. We're happy to provide support.

**Comment by jenniekim00** [comment]
Hi, not sure if it is the same visual search task as yours, but here's an available repo that uses jspsych:
https://github.com/mahiluthra/working_memory_tests

She has multiple tasks in her repo, but I think either the symmetry span task or visual array task she has might fit your needs the most. Her code is very sophisticated and well written, however it did take me a while to comprehend it fully since I am also a novice programmer. The customized plugins from her task might be a good start to get an idea on your version, since she also uses a grid for stimuli. You can find more information on the original plugin here: https://www.jspsych.org/plugins/jspsych-vsl-grid-scene/.  

I suggest searching keywords in github for similar repos and also searching in this discussion forum, with the chance that someone has done something similar. Definitely look into the the tutorials and workshops that jodeleeuw has mentioned! Hope this helps!
